accepted/future-releases/nnbd/number-operation-typing.md

```diff
@@ -62,47 +62,81 @@ T add<T extends num>(T a, T b) => a + b;
 
 because any attempt to plug in actual values will give sound results.
 
-Users also understand that `clamp` will return either the receiver or one of the arguments. If those all have the same type, the result will have that type.
+Users also understand that `clamp` will return either the receiver or one of the arguments. If those all have the same type, the result will have that type.
 
 ## Solution
 
 ### Extended Rules
 
-We extend the special-casing rules of `+`, `-`, `*` and `%` to also cover calls of the `remainder` method, and to also work with type parameters which extend `num` , `int` or `double`. Finally, if the second operand is a `double` and the first is a `num`, the result is guaranteed to be a `double`.
+We extend the special-casing rules of `+`, `-`, `*` and `%` to also cover calls of the `remainder` method, and to also work with type parameters which extend `num` , `int` or `double`. Finally, if the second operand is a `double` and the first is a `num`, the result is guaranteed to be a `double`.
 
-That is:
+Let `e` be an expression of one of the forms `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, where the static type of `e1` is a non-`Never` type *T* where *T* <: `num` and the static type of `e2` is *S*. Then:
```

Comment:
I actually want `T&double + dynamic` to have type `T&double`.  I can't do that if I pre-constrain *S* to being a subtype of `num`.

Reply:
Which is exactly the reason why I said "assignable to" and not "subtype of".

Reply:
Ah. :)

I'm still not sure why it's error prone to not make requirements on the type of the operand before deciding that it's a number operation based on the receiver. 

If this is a number operation (the receiver is a non-`Never` subtype of `num`, then we use these rules for the static type of the operation instead of using the normal rule based on the declaration of `+`, which would give `num` or `double`).

The second operand must still be assignable to num, otherwise it's a compile-time error, for all the normal reasons.

Reply:
Maybe it should be made explicit that the usual static checks apply.

Given that the type of `e1 + e2` does not follow the normal rules, it is not obvious that the static checks that we normally apply to a member invocation are applicable to `e1 + e2`. And we can't just say "Consider a program with no compile-time errors" because the new special case rules about numbers are needed in order to eliminate some errors that would otherwise exist. 

If you say "where *S* is assignable to `num`" then we'd explicitly say that this whole set of special rules don't apply otherwise, and that would make those other cases an error according to all the normal rules whenever appropriate.

Without that, I think we're saying (line 75) that `42.0 + "Hello!";` is correct and has static type `double`.

Reply:
Reasonable. Let me try to rephrase it.